lib/rgeo/cartesian/planar_graph.rb

```diff
@@ -0,0 +1,400 @@
+          he_end = incident_edges[points.last.coordinates].find { |he| he.destination == points.first }
+
+          points.each_cons(2) do |s, e|
+            edge = Segment.new(s, e)
+            if s == he_start.origin
+              he = HalfEdge.new(e)
+              he_start.twin = he
+              he.twin = he_start
+              insert_half_edge(he)
+            elsif e == he_end.origin
+              he = HalfEdge.new(s)
+              he_end.twin = he
+              he.twin = he_end
+              insert_half_edge(he)
+            else
+              create_half_edge(edge)
+            end
+            @edges << edge
+          end
+          @edges.delete(seg)
+        end
+      end
+
+      # GeometryGraph is a PlanarGraph that is built by adding
+      # geometries instead of edges. The GeometryGraph will
+      # hold a reference to an arbitrary HalfEdge on the
+      # interior of the geometry for every boundary in the geometry.
+      # For example, a polygon will have a reference to a HalfEdge for its
+      # exterior shell and one for every hole.
+      class GeometryGraph < PlanarGraph
+        # GeomEdge will be used to store the references to the HalfEdges
+        GeomEdge = Struct.new(:exterior_edge, :interior_edges)
+
+        def initialize(geom)
+          super()
+          @parent_geometry = geom
+          @geom_edges = []
+          add_geometry(geom)
+        end
+        attr_reader :parent_geometry, :geom_edges
+
+        private
+
+        # Adds a geometry to the graph and finds its
+        # reference HalfEdge(s).
+        #
+        # @param geom [RGeo::Feature::Instance]
+        def add_geometry(geom)
+          case geom
+          when Feature::Point
+            # Can't handle points yet, so just add an empty entry for them
+            @geom_edges << GeomEdge.new
+          when Feature::LineString, Feature::LinearRing
+            add_line_string(geom)
+          when Feature::Polygon
+            add_polygon(geom)
+          when Feature::GeometryCollection
+            add_collection(geom)
+          end
+        end
+
+        # Adds a LineString or LinearRing
+        # to the graph.
+        #
+        # @param geom [RGeo::Feature::LineString]
+        def add_line_string(geom)
+          # TODO: should we handle empty linestrings?
+          add_edges(geom.segments)
+
+          hedge = unless geom.is_empty?
+                    @incident_edges[geom.start_point.coordinates].first
+                  end
+
+          @geom_edges << GeomEdge.new(hedge, nil)
+        end
+
+        # Adds a Polygon to the graph.
+        #
+        # Will add each shell seperately and find a CCW half-edge
+        # for the exterior ring and a CW half-edge for interior rings
+        # since these are designated as being on the interior of the polygon.
+        #
+        # It is possible that the half-edge will be nil if a suitable one cannot
+        # be found. This indicates the polygon is likely invalid.
```

Comment:
_likely_ here is a bit scary. Is there a possibility it is valid ?

Reply:
There's no scenario I can think of where this would happen and the polygon would be valid, but I suppose that may not be the most robust test :)